kafka_checked/src/main.rs

```diff
@@ -0,0 +1,236 @@
+// https://gitlab.met.no/oda/oda/-/blob/main/internal/services/oda-kvkafka/consume-checked.go?ref_type=heads
+// https://gitlab.met.no/tjenester/oda/-/blob/dev/base/oda-kvkafka/deployment.yaml?ref_type=heads
+
+use chrono::NaiveDateTime;
+use kafka::consumer::{Consumer, FetchOffset, GroupOffsetStorage};
+use quick_xml::de::from_str;
+use serde::Deserialize;
+use std::env;
+use substring::Substring;
+use thiserror::Error;
+use tokio_postgres::NoTls;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("postgres returned an error: {0}")]
+    Database(#[from] tokio_postgres::Error),
+    #[error(
+        "no Timeseries ID found for this data - station {:?}, param {:?}",
+        station,
+        param
+    )]
+    NoTs { station: i32, param: i32 },
+}
+
+#[derive(Debug, Deserialize)]
+/// Represents <KvalobsData>...</KvalobsData>
+struct KvalobsData {
+    station: Vec<Stations>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <station>...</station>
+struct Stations {
+    #[serde(rename = "@val")]
+    val: i32,
+    typeid: Vec<Typeid>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <typeid>...</typeid>
+struct Typeid {
+    #[serde(rename = "@val")]
+    val: i32,
+    obstime: Vec<Obstime>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <obstime>...</obstime>
+struct Obstime {
+    #[serde(rename = "@val")]
+    val: String, // avoiding parsing time at this point...
+    tbtime: Vec<Tbtime>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <tbtime>...</tbtime>
+struct Tbtime {
+    #[serde(rename = "@val")]
+    val: String, // avoiding parsing time at this point...
+    sensor: Vec<Sensor>,
+    kvtextdata: Option<Vec<Kvtextdata>>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <sensor>...</sensor>
+struct Sensor {
+    #[serde(rename = "@val")]
+    val: i32,
+    level: Vec<Level>,
+}
+/// Represents <level>...</level>
+#[derive(Debug, Deserialize)]
+struct Level {
+    #[serde(rename = "@val")]
+    val: i32,
+    kvdata: Option<Vec<Kvdata>>,
+}
+/// Represents <kvdata>...</kvdata>
+#[derive(Debug, Deserialize)]
+struct Kvdata {
+    #[serde(rename = "@paramid")]
+    paramid: i32,
+    original: Option<f64>,
+    corrected: Option<f64>,
+    controlinfo: Option<String>,
+    useinfo: Option<String>,
+    cfailed: Option<String>,
+}
+/// Represents <kvtextdata>...</kvtextdata>
+#[derive(Debug, Deserialize)]
+struct Kvtextdata {
+    _paramid: Option<i32>,
+    _original: Option<String>,
+}
+#[derive(Debug, Deserialize)]
+struct Tsid {
+    station: i32,
+    paramid: i32,
+    typeid: i32,
+    sensor: i32,
+    level: i32,
```

Comment:
The structure of checked float data appears to rely on all 5 of these existing, however they do not exist for text data (sensor and level that is). So it might make sense to make them options for flexibility in that respect.

Reply:
Hmm, extra question in that case: does kvalobs create a default/zero value for data that obsinn sends without a sensor/level? If it does then we probably want to detect that and turn it into a `None`, otherwise it won't match what's in our database and you'll get "`NoTs`" errors for a bunch of timeseries that actually exist.

Reply:
Its possible, I do see a lot of lovely 0s... but that is the actual default in this case?
	<sensor val=\"0\">            
		<level val=\"0\">

Reply:
Whether it's an actual default or not, that's not what we're ingesting from obsinn, so one of the two has to change